agent/agent.go

```diff
@@ -180,22 +236,57 @@ func (a *agent) run(ctx context.Context) error {
 
 	// The startup script should only execute on the first run!
 	if oldMetadata == nil {
+		a.setLifecycle(codersdk.WorkspaceAgentLifecycleStarting)
+
+		// Perform overrides early so that Git auth can work even if users
+		// connect to a workspace that is not yet ready. We don't run this
+		// concurrently with the startup script to avoid conflicts between
+		// them.
+		if metadata.GitAuthConfigs > 0 {
+			err := gitauth.OverrideVSCodeConfigs(a.filesystem)
+			if err != nil {
+				a.logger.Warn(ctx, "failed to override vscode git auth configs", slog.Error(err))
```

Comment:
nit: should the agent quit here?

Reply:
We probably shouldn't, this could fail for a multitude of reasons, like a user setting `chmod 0000` on a folder. The worst that can happen is a degraded user experience (git auth in vscode not working). Ultimately we might want to:

- Continue executing startup script after this
- Set lifecycle to `start_error` because this failed
- Result (error message) must be visible in build/startup logs

Reply:
> The worst that can happen is a degraded user experience (git auth in vscode not working).

Yes, that's the situation I was considering. What can user do in this case? Restart the workspace until it works?

Reply:
If the change was done before, failure here doesn't really matter either (btw). Restarting will most likely not help, unless it's a problem mounting FS or similar. They'll most likely need to resolve the issue in the workspace, then restart, or create a new one.

Reply:
For now, I've only added a note. I'll revisit this behavior when I add startup log streaming 👍🏻.